tests/unit/test_network_auth.py

```diff
@@ -443,31 +552,42 @@ def check_returncode(self) -> None:
 @pytest.mark.parametrize(
     "url, expect",
     [
-        ("http://example.com/path1", (None, None)),
-        # path1 URLs will be resolved by netloc
-        ("http://user@example.com/path3", ("user", "user!netloc")),
-        ("http://user2@example.com/path3", ("user2", "user2!netloc")),
-        # path2 URLs will be resolved by index URL
-        ("http://example.com/path2/path3", (None, None)),
-        ("http://foo@example.com/path2/path3", ("foo", "foo!url")),
+        # It's not obvious, but this url ultimately resolves to index url
+        # http://example.com/path2, so we get the creds for that index.
+        ("http://example.com/path1", ("saved-user1", "pw1")),
+        ("http://saved-user1@example.com/path2", ("saved-user1", "pw1")),
+        ("http://saved-user2@example.com/path2", ("saved-user2", "pw2")),
+        ("http://new-user@example.com/path2", ("new-user", None)),
+        ("http://example.com/path2/path3", ("saved-user1", "pw1")),
+        ("http://foo@example.com/path2/path3", ("foo", None)),
     ],
 )
 def test_keyring_cli_get_password(
     monkeypatch: pytest.MonkeyPatch,
     url: str,
     expect: Tuple[Optional[str], Optional[str]],
 ) -> None:
+    keyring_subprocess = KeyringSubprocessResult()
     monkeypatch.setattr(pip._internal.network.auth.shutil, "which", lambda x: "keyring")
     monkeypatch.setattr(
-        pip._internal.network.auth.subprocess, "run", KeyringSubprocessResult()
+        pip._internal.network.auth.subprocess, "run", keyring_subprocess
     )
     auth = MultiDomainBasicAuth(
         index_urls=["http://example.com/path2", "http://example.com/path3"],
         keyring_provider="subprocess",
     )
 
-    actual = auth._get_new_credentials(url, allow_netrc=False, allow_keyring=True)
-    assert actual == expect
+    with keyring_subprocess.add_credential("example.com", "example", "!netloc"):
+        with keyring_subprocess.add_credential(
+            "http://example.com/path2/", "saved-user1", "pw1"
+        ):
+            with keyring_subprocess.add_credential(
+                "http://example.com/path2/", "saved-user2", "pw2"
+            ):
```

Comment:
Urg, so this got ugly. I was hoping to use this nicer syntax for opening multiple context managers at once, but apparently it's not supported on Python 3.8:

```suggestion
    with (
        keyring_subprocess.add_credential("example.com", "example", "!netloc"),
        keyring_subprocess.add_credential("http://example.com/path2/", "saved-user1", "pw1"),
        keyring_subprocess.add_credential("http://example.com/path2/", "saved-user2", "pw2"),
    ):
```

I could rework this `add_credential` method to instead take an array of triplets or something, but IMO that makes the api kind of messy.

I'm inclined to leave this the way it is (and clean it up when we drop support for versions of python that don't support the above syntax), but I'm happy to approach this however folks prefer.